src/test/java/org/jenkins/plugins/lockableresources/LockStepTest.java

```diff
@@ -516,6 +519,48 @@ public void evaluate() throws Throwable {
 		});
 	}
 
+	@Issue("JENKINS-40368")
+	@Test
+	public void hardKillWithWaitingRuns() throws Exception {
+		story.addStep(new Statement() {
+			@Override
+			public void evaluate() throws Throwable {
+				LockableResourcesManager.get().createResource("resource1");
+				WorkflowJob p = story.j.jenkins.createProject(WorkflowJob.class, "p");
+				p.setDefinition(new CpsFlowDefinition("retry(99) {\n" +
+						"    lock('resource1') {\n" +
+						"        semaphore('wait-inside')\n" +
+						"     }\n" +
+						"}", true));
+
+				WorkflowRun prevBuild = null;
+				for (int i = 0; i < 3; i++) {
+					WorkflowRun rNext = p.scheduleBuild2(0).waitForStart();
+					if (prevBuild != null) {
+						story.j.waitForMessage("[resource1] is locked, waiting...", rNext);
+						Executor ex = prevBuild.getExecutor();
+						assertNotNull(ex);
+						ex.interrupt();
+						story.j.waitForMessage("Click here to forcibly terminate running steps", prevBuild);
+						prevBuild.doTerm();
+						story.j.waitForMessage("Click here to forcibly kill entire build", prevBuild);
+						prevBuild.doKill();
+						story.j.waitForMessage("Hard kill!", prevBuild);
+						story.j.waitForCompletion(prevBuild);
+						story.j.assertBuildStatus(Result.ABORTED, prevBuild);
+					}
+
+					story.j.waitForMessage("Lock acquired on [resource1]", rNext);
+
+					SemaphoreStep.waitForStart("wait-inside/" + (i + 1), rNext);
+					prevBuild = rNext;
+				}
+				prevBuild.doKill();
+				story.j.assertBuildStatus(Result.ABORTED, story.j.waitForCompletion(prevBuild));
```

Comment:
Missing an important assert: a new build should succeed acquiring the lock and finish correctly after all the kill and unlock dance.